src/cryptonote_core/blockchain.cpp

```diff
@@ -1550,49 +1573,100 @@ bool Blockchain::validate_miner_transaction(
         CHECK_AND_ASSERT_MES(
                 money_in_use >= reward_parts.miner_fee, false, "base reward calculation bug");
         base_reward = money_in_use - reward_parts.miner_fee;
-    }
 
-    if (version < feature::ETH_BLS) {
-        if (b.reward >
-            reward_parts.base_miner + reward_parts.miner_fee + reward_parts.service_node_total) {
+    } else if (version == hf::hf19_reward_batching) {
```

Comment:
Are we missing HF21 here? I think so, we should patch this like so:

```cpp
} else if (version >= hf::hf19_reward_batching && hf <= hf::hf20_eth_transition) {
  // pre-existing code
} else { // NOTE: Change to fallback else branch to capture the rest.
}
```

Suggest avoiding the feature flags so that we can visually pattern match the HF numbers are all being handled correctly.

Reply:
It was indeed missing the checks for 20.  The function logic is now shuffle around quite a bit (and renamed to `validate_block_rewards` instead of `validate_miner_transaction`) because the HF21+ checks are pretty much independent of all the miner_tranactions bits needed for <=HF20.

Reply:
> Suggest avoiding the feature flags so that we can visually pattern match the HF numbers are all being handled correctly.

I agree in principle, but the fix in this case ended up splitting the logic into two versions with virtually no overlap: a "eth-bls" block, and a "non-eth-bls" block.